File: bot/kick_off.py
""""Module that handles the kickoffs"""
from rlutilities.mechanics import Dodge, AerialTurn
from rlutilities.linear_algebra import *

import math

from util import get_closest_small_pad, sign, distance_2d, lerp
from steps import Step
from custom_drive import CustomDrive as Drive


def init_kickoff(agent):
    """"Method that initializes the kickoff"""
    if abs(agent.info.my_car.location[0]) < 250:
        pad = get_closest_small_pad(agent, vec3(0, sign(agent.team) * 4608, 18))
        target = vec3(pad.location[0], pad.location[1], pad.location[2]) + sign(agent.team) * vec3(20, 0, 0)
        agent.kickoffStart = "Center"
    elif abs(agent.info.my_car.location[0]) < 1000:
        target = vec3(0.0, sign(agent.team) * 2816.0, 70.0) + sign(agent.team) * vec3(0, 300, 0)
        agent.kickoffStart = "offCenter"
    else:
        target = agent.info.my_car.location + 300 * agent.info.my_car.forward()
        agent.kickoffStart = "Diagonal"
    agent.drive = Drive(agent.info.my_car)
    agent.drive.target = target
    agent.drive.speed = 2400
    agent.step = Step.Drive
    agent.drive.step(agent.fps)
    agent.controls = agent.drive.controls


def kick_off(agent):
    ball = agent.info.ball
    car = agent.info.my_car
    t = distance_2d(ball.location, car.location) / 2200
    batmobile_resting = 18.65
    robbies_constant = (ball.location - vec3(0, 0,
                                                        92.75 - batmobile_resting) - car.location - car.velocity * t) * 2 * t ** -2
    robbies_boost_constant = dot(normalize(xy(car.forward())), normalize(xy(robbies_constant))) > (
                0.3 if car.on_ground else 0.1)
    """"Module that performs the kickoffs"""
    if agent.kickoffStart == "Diagonal":
        if agent.step is Step.Drive:
            agent.drive.step(agent.fps)
            agent.controls = agent.drive.controls
            if agent.drive.finished:
                ball_location = ball.location + vec3(0, -sign(agent.team) * 500, 0)
                target = car.location + 250 * normalize(ball_location - car.location)
                agent.drive = Drive(car)
                agent.drive.target = target
                agent.drive.speed = 2400
                agent.step = Step.Drive_1
        if agent.step is Step.Drive_1:
            agent.drive.step(agent.fps)
            agent.controls = agent.drive.controls
            if agent.drive.finished:
                target = vec3(dot(rotation(math.radians(-sign(agent.info.team) * sign(car.location[0]) * 60)),
                                  vec2(car.forward())) * 10000)
                preorientation = dot(
                    axis_to_rotation(vec3(0, 0, math.radians(-sign(agent.info.team) * -sign(car.location[0]) * 30))),
                    car.rotation)
                setup_first_dodge(agent, 0.05, 0.3, target, preorientation)
        elif agent.step is Step.Dodge_1:
            agent.timer += agent.info.time_delta
            if agent.timer > 0.8:
                lerp_var = lerp(normalize(robbies_constant), normalize(
                    ball.location - vec3(0, 0, 92.75 - batmobile_resting) - car.location),
                                0.8)
                agent.turn.target = look_at(lerp_var, vec3(0, 0, 1))
                agent.turn.step(agent.info.time_delta)
                agent.controls = agent.turn.controls
                if car.on_ground:
                    agent.step = Step.Catching
            else:
                agent.dodge.step(agent.info.time_delta)
                agent.controls = agent.dodge.controls
            agent.controls.boost = robbies_boost_constant
    elif agent.kickoffStart == "Center":
        if agent.step is Step.Drive:
            agent.drive.step(agent.info.time_delta)
            agent.controls = agent.drive.controls
            if agent.drive.finished:
                target = vec3(dot(rotation(math.radians(-65)), vec2(car.forward())) * 10000)
                preorientation = dot(axis_to_rotation(vec3(0, 0, math.radians(45))),
                                     car.rotation)
                setup_first_dodge(agent, 0.05, 0.4, target, preorientation)
        elif agent.step is Step.Dodge_1:
            agent.timer += agent.info.time_delta
            if agent.timer > 0.8:
                agent.turn.target = look_at(xy(ball.location - car.location), vec3(0, 0, 1))
                agent.turn.step(agent.info.time_delta)
                agent.controls = agent.turn.controls
                set_steer(agent)
            else:
                agent.dodge.step(agent.info.time_delta)
                agent.controls = agent.dodge.controls
            agent.controls.boost = robbies_boost_constant
        elif agent.step is Step.Steer:
            agent.drive.step(agent.info.time_delta)
            agent.controls = agent.drive.controls
            if distance_2d(car.location, ball.location) < 800:
                agent.step = Step.Dodge_2
                agent.dodge = Dodge(car)
                agent.dodge.duration = 0.075
                agent.dodge.target = ball.location
        elif agent.step is Step.Dodge_2:
            agent.dodge.step(agent.info.time_delta)
            agent.controls = agent.dodge.controls
            if agent.dodge.finished and car.on_ground:
                agent.step = Step.Catching
    elif agent.kickoffStart == "offCenter":
        if agent.step is Step.Drive:
            agent.drive.step(agent.fps)
            agent.controls = agent.drive.controls
            if distance_2d(car.location, agent.drive.target) < 650:
                target = vec3(dot(rotation(math.radians(-sign(agent.info.team) * -sign(car.location[0]) * 100)),
                                  vec2(car.forward())) * 10000)
                preorientation = dot(
                    axis_to_rotation(vec3(0, 0, math.radians(-sign(agent.info.team) * sign(car.location[0]) * 30))),
                    car.rotation)
                setup_first_dodge(agent, 0.05, 0.4, target, preorientation)
        elif agent.step is Step.Dodge_1:
            agent.timer += agent.info.time_delta
            if agent.timer > 0.8:
                lerp_var = lerp(normalize(robbies_constant), normalize(
                    ball.location - vec3(0, 0, 92.75 - batmobile_resting) - car.location),
                                0.25)
                agent.turn.target = look_at(lerp_var, vec3(0, 0, 1))
                agent.turn.step(agent.info.time_delta)
                agent.controls = agent.turn.controls
                set_steer(agent)
            else:
                agent.dodge.step(agent.info.time_delta)
                agent.controls = agent.dodge.controls
            agent.controls.boost = robbies_boost_constant
        elif agent.step is Step.Steer:
            agent.drive.step(agent.fps)
            agent.controls = agent.drive.controls
            if distance_2d(ball.location, car.location) < 800:
                agent.step = Step.Dodge_2
                agent.dodge = Dodge(car)
                agent.dodge.duration = 0.075
                agent.dodge.target = ball.location
        elif agent.step is Step.Dodge_2:
            agent.dodge.step(agent.fps)
            agent.controls = agent.dodge.controls
            if agent.dodge.finished and car.on_ground:
                agent.step = Step.Catching


def set_steer(agent):
    if agent.info.my_car.on_ground:
        agent.step = Step.Steer
        target = agent.info.ball.location
        agent.drive = Drive(agent.info.my_car)
        agent.drive.target = target
        agent.drive.speed = 2400


def setup_first_dodge(agent, duration, delay, target, preorientation):
    agent.dodge = Dodge(agent.info.my_car)
    agent.turn = AerialTurn(agent.info.my_car)
    agent.dodge.duration = duration
    agent.dodge.delay = delay
    agent.dodge.target = target
    agent.dodge.preorientation = preorientation
    agent.timer = 0.0
    agent.step = Step.Dodge_1

File: NV-Derevo-CPP/NV-Derevo/src/kickoff.cc
#include "kickoff.h"
#include "utils.h"


void InitKickoff(NVDerevo & derevo, Game & game) {
    vec3 target;
    if (abs(game.my_car->position[0]) < 250) {
        target = vec3{sign(derevo.team) * 20.0f, sign(derevo.team) * 4240.0f, 6.0f};
        derevo.kickoff_start = Center;
    } else if (abs(game.my_car->position[0]) < 1000) {
        target = vec3{0.0f, sign(derevo.team) * (2816.0f + 300.0f), 70.0f};
        derevo.kickoff_start = OffCenter;
    } else {
        target = game.my_car->position + 300 * game.my_car->forward();
        derevo.kickoff_start = Diagonal;
    }
    derevo.drive = new Drive(*game.my_car);
    derevo.drive->target = target;
    derevo.drive->speed = 2400;
    derevo.step = Driving;
    derevo.drive->step(game.time_delta);
    derevo.controls = derevo.drive->controls;    
}

void KickOff(NVDerevo & derevo, Game & game) {
    Ball ball = game.ball;
    Car car = *game.my_car;
    float t = Distance2D(ball.position, car.position) / 2200;
    float batmobile_resting = 18.65f;
    vec3 robbies_constant = (ball.position - vec3{0, 0, 92.75f - batmobile_resting} - car.position - car.velocity * t) * 2 * std::pow(t, -2);
    bool robbies_boost_constant = dot(normalize(xy(car.forward())), normalize(xy(robbies_constant))) > (car.on_ground ? 0.3 : 0.1);
    // Module that performs the kickoffs
    std::cout << derevo.step << std::endl;
    if (derevo.kickoff_start == Diagonal) {
        if (derevo.step == Driving) {
            derevo.drive->step(game.time_delta);
            derevo.controls = derevo.drive->controls;
            if (derevo.drive->finished) {
                vec3 ball_location = ball.position + vec3{0, -sign(derevo.team) * 500.0f, 0};
                vec3 target = car.position + 250 * normalize(ball_location - car.position);
                derevo.drive = new Drive(*game.my_car);
                derevo.drive->target = target;
                derevo.drive->speed = 2400;
                derevo.step = Driving_1;
            }
        } else if (derevo.step == Driving_1) {
            derevo.drive->step(game.time_delta);
            derevo.controls = derevo.drive->controls;
            if (derevo.drive->finished) {
                float target_degrees = -sign(derevo.team) * sign(car.position[0]) * 60.0f;
                float target_radians = (atan(1.0f)* 4.0f * target_degrees) / 180.0f;
                vec3 target = vec3(dot(rotation(target_radians), vec2(car.forward())) * 10000.0f);
                float rot_degrees = -sign(derevo.team) * -sign(car.position[0]) * 30.0f;
                float rot_radians = (atan(1.0f)* 4.0f * rot_degrees) / 180.0f;
                mat3 preorientation = dot(axis_to_rotation(vec3{0, 0, rot_radians}), car.orientation);
                SetupFirstDodge(derevo, game, 0.05, 0.3, target, preorientation);
            }
        } else if (derevo.step == Dodging_1) {
            if (derevo.dodge->timer > 0.8) {
                vec3 lerp_var = lerp(normalize(robbies_constant), normalize(ball.position - vec3{0.0f, 0.0f, 92.75f - batmobile_resting} - car.position), 0.8f);
                derevo.turn->target_orientation = look_at(lerp_var, vec3{0, 0, 1});
                derevo.turn->step(game.time_delta);
                derevo.controls = derevo.turn->controls;
                if (car.on_ground) {
                    derevo.step = Shooting;
                }
            } else {
                derevo.dodge->step(game.time_delta);
                derevo.controls = derevo.dodge->controls;
            }
            derevo.controls.boost = robbies_boost_constant;
        }
    } else if (derevo.kickoff_start == Center) {
        if (derevo.step == Driving) {
            derevo.drive->step(game.time_delta);
            derevo.controls = derevo.drive->controls;
            if (derevo.drive->finished) {
                float target_degrees = -65.0f;
                float target_radians = (atan(1.0f)* 4.0f * target_degrees) / 180.0f;
                vec3 target = vec3(dot(rotation(target_radians), vec2{car.forward()}) * 10000.0f);
                float rot_degrees = 45.0f;
                float rot_radians = (atan(1.0f)* 4.0f * rot_degrees) / 180.0f;
                mat3 preorientation = dot(axis_to_rotation(vec3{0, 0, rot_radians}), car.orientation);
                SetupFirstDodge(derevo, game, 0.05, 0.4, target, preorientation);
            }
        } else if (derevo.step == Dodging_1) {
            if (derevo.dodge->timer > 0.8) {
                derevo.turn->target_orientation = look_at(xy(ball.position - car.position), vec3{0, 0, 1});
                derevo.turn->step(game.time_delta);
                derevo.controls = derevo.turn->controls;
                SetSteer(derevo, game);
            } else {
                derevo.dodge->step(game.time_delta);
                derevo.controls = derevo.dodge->controls;
            }
            derevo.controls.boost = robbies_boost_constant;
        } else if (derevo.step == Steering) {
            derevo.drive->step(game.time_delta);
            derevo.controls = derevo.drive->controls;
            if (Distance2D(car.position, ball.position) < 800) {
                derevo.dodge = new Dodge(*game.my_car);
                derevo.dodge->duration = 0.075;
                derevo.dodge->target = ball.position;
                derevo.step = Dodging_2;
            }
        } else if (derevo.step == Dodging_2) {
            derevo.dodge->step(game.time_delta);
            derevo.controls = derevo.dodge->controls;
            if (derevo.dodge->finished && car.on_ground) {
                derevo.step = Shooting;
            }
        }
    } else if (derevo.kickoff_start == OffCenter) {
        if (derevo.step == Driving) {
            derevo.drive->step(game.time_delta);
            derevo.controls = derevo.drive->controls;
            if (Distance2D(car.position, derevo.drive->target) < 650) {
                float target_degrees = -sign(derevo.team) * -sign(car.position[0]) * 100.0f;
                float target_radians = (atan(1.0f)* 4.0f * target_degrees) / 180.0f;
                vec3 target = vec3(dot(rotation(target_radians), vec2(car.forward())) * 10000.0f);
                float rot_degrees = -sign(derevo.team) * sign(car.position[0]) * 30.0f;
                float rot_radians = (atan(1.0f)* 4.0f * rot_degrees) / 180.0f;
                mat3 preorientation = dot(axis_to_rotation(vec3{0, 0, rot_radians}), car.orientation);
                std::cout << "CPP: " << game.my_car->position << std::endl;
                std::cout << "CPP: " << (atan(1.0f)* 4.0f * target_degrees) / 180.0f << std::endl;
                SetupFirstDodge(derevo, game, 0.05, 0.4, target, preorientation);
            }
        }  else if (derevo.step == Dodging_1) {
            if (derevo.dodge->timer > 0.8) {
                vec3 lerp_var = lerp(normalize(robbies_constant), normalize(ball.position - vec3{0.0f, 0.0f, 92.75f - batmobile_resting} - car.position), 0.25f);
                derevo.turn->target_orientation = look_at(lerp_var, vec3{0, 0, 1});
                derevo.turn->step(game.time_delta);
                derevo.controls = derevo.turn->controls;
                SetSteer(derevo, game);
            } else {
                derevo.dodge->step(game.time_delta);
                derevo.controls = derevo.dodge->controls;
            }
            derevo.controls.boost = robbies_boost_constant;
        } else if (derevo.step == Steering) {
            derevo.drive->step(game.time_delta);
            derevo.controls = derevo.drive->controls;
            if (Distance2D(ball.position, car.position) < 800) {
                derevo.dodge = new Dodge(*game.my_car);
                derevo.dodge->duration = 0.075;
                derevo.dodge->target = ball.position;
                derevo.step = Dodging_2;
            }
        } else if (derevo.step == Dodging_2) {
            derevo.dodge->step(game.time_delta);
            derevo.controls = derevo.dodge->controls;
            if (derevo.dodge->finished && car.on_ground) {
                derevo.step = Shooting;
            }
        }
    }
}

void SetSteer(NVDerevo & derevo, Game & game) {
    if (game.my_car->on_ground) {
        derevo.step = Steering;
        derevo.drive = new Drive(*game.my_car);
        derevo.drive->target = game.ball.position;
        derevo.drive->speed = 2400;
    }
}

void SetupFirstDodge(NVDerevo & derevo, Game & game, float duration, float delay, vec3 target, mat3 preorientation) {
    derevo.dodge = new Dodge(*game.my_car);
    derevo.turn = new Reorient(*game.my_car);
    derevo.dodge->duration = duration;
    derevo.dodge->delay = delay;
    derevo.dodge->target = target;
    derevo.dodge->preorientation = preorientation;
    derevo.step = Dodging_1;
}